### PR TITLE
Add edit history to web UI

### PR DESCRIPTION
**Pull request by Gargron**
Follow-up to #16697

In detailed view, posts show the timestamp that they were last edited, which can be clicked to load the edit history:

image

Each entry can be clicked to display that version in a modal window:

image

The PR includes adjustments to the `GET /api/v1/statuses/:id/history` API:

- Replace `text` with `content` for consistency with remote posts
- Add `emojis` to allow properly rendering the text
- Add `account`

A refactor of the dropdown menu component was necessary to enable reusing dropdowns for more than just action menus.
___

Funded through the BMBF (BMBF Förderkennzeichen: 01IS21S29)

### CODE REVIEW

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/javascript/mastodon/components/dropdown_menu.js [hunk 1]
```diff
@@ -66,7 +74,7 @@ class DropdownMenu extends React.PureComponent {
   }
 
   handleKeyDown = e => {
-    const items = Array.from(this.node.getElementsByTagName('a'));
+    const items = Array.from(this.node.querySelectorAll('a, button'));
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/javascript/mastodon/components/dropdown_menu.js:78]: For menu items that should open in new tab, using `<a>` is a lot more natural than using `<button>`.

File: app/javascript/mastodon/components/dropdown_menu.js [hunk 2]
```diff
@@ -170,18 +178,24 @@ export default class Dropdown extends React.PureComponent {
   };
 
   static propTypes = {
-    icon: PropTypes.string.isRequired,
-    items: PropTypes.array.isRequired,
-    size: PropTypes.number.isRequired,
+    children: PropTypes.node,
+    icon: PropTypes.string,
+    items: PropTypes.oneOfType([PropTypes.array, ImmutablePropTypes.list]).isRequired,
+    loading: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/javascript/mastodon/components/dropdown_menu.js:191]: Now it is.

File: app/javascript/mastodon/components/edited_timestamp/containers/dropdown_menu_container.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { connect } from 'react-redux';
+import { openDropdownMenu, closeDropdownMenu } from 'mastodon/actions/dropdown_menu';
+import { fetchHistory } from 'mastodon/actions/history';
+import DropdownMenu from 'mastodon/components/dropdown_menu';
+
+const mapStateToProps = state => ({
+  dropdownPlacement: state.getIn(['dropdown_menu', 'placement']),
+  openDropdownId: state.getIn(['dropdown_menu', 'openId']),
+  openedViaKeyboard: state.getIn(['dropdown_menu', 'keyboard']),
+  items: state.getIn(['history', 'items']),
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/javascript/mastodon/components/edited_timestamp/containers/dropdown_menu_container.js:10]: When `HISTORY_FETCH_REQUEST` starts, the previous items are reset. This happens in a dispatch before the dropdown opens.

**Comment by Gargron** [comment]
PR now includes a much better-looking circular loading indicator.

File: app/javascript/mastodon/components/edited_timestamp/containers/dropdown_menu_container.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/javascript/mastodon/components/edited_timestamp/containers/dropdown_menu_container.js:10]: Yes, but the API reply might come after the user opened another drop-down, and there's no guarantee two API requests will return in the order they were issued.

**Comment by Gargron** [comment]
History items are now stored in a map segmented by status ID to avoid dropdowns showing wrong items due to race conditions.

File: app/javascript/mastodon/components/edited_timestamp/containers/dropdown_menu_container.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { connect } from 'react-redux';
+import { openDropdownMenu, closeDropdownMenu } from 'mastodon/actions/dropdown_menu';
+import { fetchHistory } from 'mastodon/actions/history';
+import DropdownMenu from 'mastodon/components/dropdown_menu';
+
+const mapStateToProps = (state, { statusId }) => ({
+  dropdownPlacement: state.getIn(['dropdown_menu', 'placement']),
+  openDropdownId: state.getIn(['dropdown_menu', 'openId']),
+  openedViaKeyboard: state.getIn(['dropdown_menu', 'keyboard']),
+  items: state.getIn(['history', statusId, 'items']),
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/edited_timestamp/containers/dropdown_menu_container.js:10]: Missing prop: `loading: state.getIn(['history', statusId, 'loading']),`

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Apart from the missing `loading` prop, there are some issues that probably should be addressed in later PRs:
- redux history state might grow large over time, it probably needs pruning in some way
- overall, I dislike how complicated `DropdownMenu` is, but I have no concrete step to offer you to improve that
- the overall style could need some improvement, especially the position of the dropdown menu

https://user-images.githubusercontent.com/384364/152955001-41001ded-321e-49aa-a6c5-d7740d7155e5.mp4

EDITED: just noticed it also doesn't mention whether the attachments have changed

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**